driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java

```diff
@@ -177,43 +169,23 @@ protected EncodingMetadata encodeMessageBodyWithMetadata(final BsonOutput bsonOu
             // Write the flag bits
             bsonOutput.writeInt32(flagPosition, getOpMsgFlagBits());
         } else {
-            bsonOutput.writeInt32(getOpQueryFlagBits());
+            bsonOutput.writeInt32(0);
             bsonOutput.writeCString(namespace.getFullName());
             bsonOutput.writeInt32(0);
             bsonOutput.writeInt32(-1);
 
             commandStartPosition = bsonOutput.getPosition();
 
-            if (payload == null) {
-                List<BsonElement> elements = null;
-                if (serverApi != null) {
-                    elements = new ArrayList<>(3);
-                    addServerApiElements(elements);
-                }
-                addDocument(getCommandToEncode(), bsonOutput, commandFieldNameValidator, elements);
-            } else {
-                // We're not concerned with adding ServerApi elements here.  The only reason we do it for OP_QUERY-based commands is that
-                // OP_QUERY is always used for the handshake, and we have to pass ServerApi elements in the handshake.  Other than that,
-                // all servers that support ServerApi also support OP_MSG, so this code path should never be hit.
-                addDocumentWithPayload(bsonOutput, messageStartPosition);
+            List<BsonElement> elements = null;
+            if (serverApi != null) {
+                elements = new ArrayList<>(3);
+                addServerApiElements(elements);
             }
+            addDocument(command, bsonOutput, commandFieldNameValidator, elements);
         }
         return new EncodingMetadata(commandStartPosition);
     }
 
-    private FieldNameValidator getPayloadArrayFieldNameValidator() {
-        Map<String, FieldNameValidator> rootMap = new HashMap<String, FieldNameValidator>();
-        rootMap.put(payload.getPayloadName(), payloadFieldNameValidator);
-        return new MappedFieldNameValidator(commandFieldNameValidator, rootMap);
-    }
-
-    private void addDocumentWithPayload(final BsonOutput bsonOutput, final int messageStartPosition) {
-        BsonBinaryWriter bsonBinaryWriter = new BsonBinaryWriter(bsonOutput, getPayloadArrayFieldNameValidator());
-        BsonWriter bsonWriter = new SplittablePayloadBsonWriter(bsonBinaryWriter, bsonOutput, messageStartPosition, getSettings(), payload);
-        BsonDocument commandToEncode = getCommandToEncode();
-        getCodec(commandToEncode).encode(bsonWriter, commandToEncode, EncoderContext.builder().build());
-    }
-
     private int getOpMsgFlagBits() {
         return getOpMsgResponseExpectedFlagBit();
```

Comment:
Nit - getOpMsgResponseExpectedFlagBit() is only used here - no need for two methods.

Reply:
Inlined